ESLint consolidation

Antonio-Laguna opened the pull request:
### Description of the Change

This started as let's see what's available to do. Realized there were issues from dependabot that were failing Eslint and then giving an opinionated go to that.

These are the screenshots that showcase the issues:

![Screenshot 2022-08-08 at 10 55 45](https://user-images.githubusercontent.com/946645/183388687-edab2ead-4adb-4739-9955-59a79ed774ad.png)

![Screenshot 2022-08-08 at 10 55 54](https://user-images.githubusercontent.com/946645/183388676-9f7b2b3f-5276-428b-8189-857738002954.png)

Now a quick summary of what this does:

* Created a consolidated action that runs `lint` and `test` on every workspace that has it available.
* Added `lint` to every single package that didn't have it and also linted every file.
* Removed remnants of Node 12.
* Ensured license is consistent
* Separated tests on `eslint-config` to test each restricted syntax in isolation. The previous test had a bunch of `eslint-disable` modifiers plus tests only check that the files fail linting but not why. These files _only_ fail on the intended.
* Streamlined dependencies for `eslint-config`
* Removing husky and lint-staged: This goes in tandem with the `lint` in every package. Global `eslint` does seem like a good idea but not really since there's a mixture of packages with different linting needs. Each package now lints itself if available (all are now) and we get direct feedback on each PR from the linter. 
It was also not working correctly (on my tests) since either was too greedy that it picked `node_modules` or didn't lint enough files. Using this approach gave me a quick status check on every public package.

I **have not** added any changeset since I wanted to check this with the team first. I don't think there's a need to release anything as these are just chores. Though perhaps it'd be good to release `eslint-config` as `minor` to get the `peerDependency` updates.

### Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] I have read the [**CONTRIBUTING**](https://github.com/10up/10up-toolkit/blob/develop/CONTRIBUTING.md) document.
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have added tests to cover my change.
- [X] All new and existing tests passed.
- [ ] I have added a changeset to my PR. See [**CONTRIBUTING**](https://github.com/10up/10up-toolkit/blob/develop/CONTRIBUTING.md) document for instructions

Commit:
Updating wp-env to avoid security issues

Commit:
Normalizing license

Commit:
Adding lint

Commit:
Better tests for eslint-config

Commit:
Simpler and more readable

Commit:
Linting

Commit:
Better dependencies

It was inconsistent the fact you needed to install some of the dependencies but not all. You get all of React even if you’re just linting JavaScript but you’re only requested to install WP and Jest rules

Commit:
Linting

Commit:
Removing root eslint, letting each package lint itself

Commit:
Node 12 is EOL

Commit:
Letting each package lint itself

Commit:
Running everything at once

Antonio-Laguna commented:
The new tests are also good since not all features seem OK in every OS

Commit:
Jest for babel-preset-default

Commit:
Toolkit too needs Jest

Commit:
Node 18 is not ready for primetime on some libs

Commit:
More Jest support

Commit:
Better `from` for Windows?

Commit:
Avoid duplicating work

nicholasio commented:
Can we keep lint-staged and husky and have it trigger the individual tests on each package? I really like the idea of blocking commits if things fails to lint.

nicholasio commented:
Did you check the failures on ubuntu and windows?

nicholasio commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,77 @@
+        with:
+          node-version: ${{ matrix.node }}
+
+      - name: env
```
What's this doing?

Antonio-Laguna replied:
Building up env variables for the restore keys below

nicholasio commented on packages/eslint-config/package.json:
```diff
@@ -32,31 +32,20 @@
   "peerDependencies": {
     "@babel/core": "^7.16.0",
     "@babel/eslint-parser": "^7.16.3",
+	"@wordpress/eslint-plugin": "^12.0.0",
```
Can we make this an optional dependency? We don't need it for non-wp projects and I'd rather not install WP specific things for non-wp projects;

Antonio-Laguna replied:
We can, but we install JSX, React and React and React Hooks dependencies even if we don't use them. 

I think being inconsistent and also having mixed approaches depending on what you need to use felt a bit odd. This is a dev dependency in any case. 

Let me know your thoughts !

nicholasio replied:
We don't install react, that's responsibility of the consumer. When using toolkit in WP context, the WP bundled react will be used.

toolkit already install `@wordpress/eslint-plugin` and everyone using eslint-config for WP would prob just use toolkit instead and would almost never install eslint-config directly, so leaving our eslint-config with no wp-specific stuff by default sounds reasonable to me.

Antonio-Laguna replied:
What I mean is that JSX and Hooks and React eslint packages are installed

Antonio-Laguna replied:
IMHO it's a bit strange to be asked to install other packages. Adding to that is the fact that different linting scenarios are treated differently which is why I did that.

Antonio-Laguna commented:
> Can we keep lint-staged and husky and have it trigger the individual tests on each package? I really like the idea of blocking commits if things fails to lint.

We could hook `lint-staged` to the whole linting but that feels odd. If I haven't changed those files I don't want to be blocked. At the same time, I think it makes sense for you to actually be able to commit things that aren't still 100% or that purposely fail to test things out (in any scenario).

Given you can't lint-stage per package's config (easily) I thought it makes the most sense to get a clean slate and be warned at a PR level so everyone in the PR process knows if something is suddenly not linting.  Especially when the packages had different linting issues that weren't caught before. 

In any case, this was meant as a discussion starter rather than "let's just roll with this".

>  Did you check the failures on ubuntu and windows?

I have, not sure yet what they are since the test seems to enter some sort of loop and has to be killed. Will investigate further.

Commit:
Leveraging editorconfig for indentation

Commit:
Trying to fix infinite loop

Commit:
Testing something

Commit:
Reverting

Commit:
Centralizing

nicholasio commented:
> We could hook `lint-staged` to the whole linting but that feels odd. If I haven't changed those files I don't want to be blocked. At the same time, I think it makes sense for you to actually be able to commit things that aren't still 100% or that purposely fail to test things out (in any scenario).

lint-staged would only lint the changed files.

> Given you can't lint-stage per package's config (easily) I thought it makes the most sense to get a clean slate and be warned at a PR level so everyone in the PR process knows if something is suddenly not linting. Especially when the packages had different linting issues that weren't caught before.

It's a valid approach but we've been historically using husky and lint-staged to get linting feedback at commit level. One downside of having to wait for PR bots is that it might take a while, with lint-staged you can near instant feedback at commit time and you avoid linting back-and-forth in the PR.

Antonio-Laguna commented:
> lint-staged would only lint the changed files.

Yes but it's not able to apply different linting scenarios per file which is why I changed. That's the main reason I changed that.

> It's a valid approach but we've been historically using husky and lint-staged to get linting feedback at commit level. One downside of having to wait for PR bots is that it might take a while, with lint-staged you can near instant feedback at commit time and you avoid linting back-and-forth in the PR.

It's also a valid approach to use husky and lint-staged. I'm not saying it isn't and if it wasn't because it's not doable to have lint-staged with different scenarios (global eslint wasn't working) I wouldn't have changed it. Also, we still got unlinted files somehow within the repo which, somehow, escaped that instant feedback at commit time.

In any case, if you can think of other solutions let me know as I'm more than happy to take another path. I feel like lint-staged is not the right fit for monorepos with different needs and different packages but I could be wrong.

Commit:
Finishing linting

Commit:
Will this make any difference?

Commit:
Testing windows again

Commit:
More logging

Commit:
Not working

Commit:
Removing

Commit:
Merge branch 'develop' into feature/eslint-consolidation

# Conflicts:
#	package-lock.json
#	projects/10up-theme/package.json

Commit:
Fixing usage of loadFormatter

`loadFormatter` returns a promise which, once resolved, returns an object with a `format` function

Commit:
Do not ignore scripts

nicholasio commented:
> In any case, if you can think of other solutions let me know as I'm more than happy to take another path. I feel like lint-staged is not the right fit for monorepos with different needs and different packages but I could be wrong.

fair enough.

Looks like the test on windows is still failling/running indefinatelly, once that's resolved I think we should be good to merge this.

Antonio-Laguna commented:
I’m not sure if the tests will pass, will try to, at least, make them fail. I got a virtual box instance of W11 and was able to move a bit forward but not having windows makes it complicated 😅

Will keep you posted!

Commit:
Refactoring a bit to appease windows

Commit:
Fixing files

Commit:
Can’t really rename, breaking change

Antonio-Laguna commented:
Windows is failing because of this issue https://github.com/npm/cli/issues/5454 

I'll try to come up with some workaround :/

Commit:
Using exports

This leverages exports which Eslint seems to be able to leverage to overcome this issue without creating a breaking change:

https://github.com/npm/cli/issues/5454

This is all in a single commit to ease revert if needed.

Commit:
Adding gitattributes

Commit:
Normalizing paths to POSIX

Commit:
Hopefully fixing

Commit:
Fixing name

Antonio-Laguna commented:
After 40 commits, countless hours, and (20gb) windows virtual machine this seems to be all green now.

cc @fabiankaegy @nicholasio

nicholasio commented on packages/eslint-config/package.json:
```diff
@@ -2,65 +2,62 @@
+	"test:wordpress": "node fixtures/test-eslint-config.js --config-file=wordpress",
+	"test:node": "node fixtures/test-eslint-config.js --config-file=nodejs"
+  },
+  "exports": {
```
Would this require a major bump?

Antonio-Laguna replied:
No.

This was a major concern. I didn't want to create a breaking change. Renaming node to nodejs would have been breaking so I tried with exports to see if it worked. 

The exports change just required changes on some internal routes but no changes on external package usages.

nicholasio committed:
make  optional and update README.md for eslint config

nicholasio reviewed (approved):
@Antonio-Laguna I made one small update to make `@wordpress/eslint-plugin` optional. I did that since we also use the eslint-config on non-wp projects and it's how it was set up from the start. I agree we can make the other deps non-optional. 

For wp projects we use toolklit which already install the eslint-plugin.

Thanks for all your work on this!

Antonio-Laguna commented:
That's a sensible addition. Thanks for doing that :)
